File: app/src/main/java/com/gnoemes/shikimoriapp/presentation/view/alternative/episodes/adapter/EpisodeAlternativeOptionAdapterDelegate.java
package com.gnoemes.shikimoriapp.presentation.view.alternative.episodes.adapter;

import android.graphics.drawable.Drawable;
import android.support.annotation.NonNull;
import android.support.constraint.ConstraintLayout;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.gnoemes.shikimoriapp.R;
import com.gnoemes.shikimoriapp.entity.anime.series.presentation.EpisodeOptionAction;
import com.gnoemes.shikimoriapp.entity.anime.series.presentation.EpisodeOptionsItem;
import com.gnoemes.shikimoriapp.entity.app.presentation.BaseItem;
import com.gnoemes.shikimoriapp.utils.view.DrawableHelper;
import com.hannesdorfmann.adapterdelegates3.AdapterDelegate;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

public class EpisodeAlternativeOptionAdapterDelegate extends AdapterDelegate<List<BaseItem>> {

    private EpisodeOptionCallback callback;

    EpisodeAlternativeOptionAdapterDelegate(@NonNull EpisodeOptionCallback callback) {
        this.callback = callback;
    }

    @Override
    protected boolean isForViewType(@NonNull List<BaseItem> items, int position) {
        return items.get(position) instanceof EpisodeOptionsItem;
    }

    @NonNull
    @Override
    protected RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.item_episode_options, parent, false);
        return new ViewHolder(view);
    }

    @Override
    protected void onBindViewHolder(@NonNull List<BaseItem> items, int position, @NonNull RecyclerView.ViewHolder holder, @NonNull List<Object> payloads) {
        EpisodeOptionsItem optionsItem = (EpisodeOptionsItem) items.get(position);
        ViewHolder viewHolder = (ViewHolder) holder;
        viewHolder.bind(optionsItem);
    }

    class ViewHolder extends RecyclerView.ViewHolder {

        @BindView(R.id.constraint)
        ConstraintLayout layout;

        @BindView(R.id.btn_continue)
        TextView continueBtn;

        @BindView(R.id.text_episode)
        TextView nextEpisode;

        @BindView(R.id.image_alternative)
        ImageView alternativeBtn;

        public ViewHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);

            Drawable card = DrawableHelper
                    .withContext(itemView.getContext())
                    .withDrawable(R.drawable.card)
                    .withAttributeColor(R.attr.colorBackgroundWindow)
                    .tint()
                    .get();

            layout.setBackground(card);

            Drawable icon = DrawableHelper
                    .withContext(itemView.getContext())
                    .withDrawable(R.drawable.ic_play_circle_filled)
                    .withAttributeColor(R.attr.colorText)
                    .tint()
                    .get();

            continueBtn.setCompoundDrawablesWithIntrinsicBounds(icon, null, null, null);

        }

        public void bind(EpisodeOptionsItem optionsItem) {

            String buttonText;
            if (optionsItem.isFirst()) {
                buttonText = itemView.getResources().getString(R.string.watch_now);
            } else {
                buttonText = itemView.getResources().getString(R.string.watch_continue);

            }
            nextEpisode.setText(optionsItem.getEpisodeItem().getEpisodeFull());

            continueBtn.setText(buttonText);

            continueBtn.setOnClickListener(v -> callback.onAction(EpisodeOptionAction.ALTERNATIVE_SOURCE, optionsItem.getEpisodeItem()));
            alternativeBtn.setVisibility(View.GONE);
        }
    }
}

File: app/src/main/java/com/gnoemes/shikimoriapp/presentation/view/alternative/episodes/adapter/EpisodeAlternativeOptionAdapterDelegate.kt
package com.gnoemes.shikimoriapp.presentation.view.alternative.episodes.adapter

import android.annotation.SuppressLint
import android.support.v7.widget.RecyclerView
import android.view.View
import android.view.ViewGroup
import com.gnoemes.shikimoriapp.R
import com.gnoemes.shikimoriapp.entity.anime.series.presentation.EpisodeItem
import com.gnoemes.shikimoriapp.entity.anime.series.presentation.EpisodeOptionAction
import com.gnoemes.shikimoriapp.entity.anime.series.presentation.EpisodeOptionsItem
import com.gnoemes.shikimoriapp.entity.app.presentation.BaseItem
import com.gnoemes.shikimoriapp.utils.inflate
import com.gnoemes.shikimoriapp.utils.themeDrawable
import com.hannesdorfmann.adapterdelegates3.AdapterDelegate
import kotlinx.android.synthetic.main.item_episode.view.*
import kotlinx.android.synthetic.main.item_episode_options.view.*

class EpisodeAlternativeOptionAdapterDelegate(val callback: ((EpisodeOptionAction, EpisodeItem) -> Unit)
) : AdapterDelegate<List<BaseItem>>() {

    override fun onCreateViewHolder(parent: ViewGroup?): RecyclerView.ViewHolder =
            ViewHolder(parent!!.inflate(R.layout.item_episode_options, false))

    override fun isForViewType(items: List<BaseItem>, position: Int): Boolean =
            items[position] is EpisodeOptionsItem

    override fun onBindViewHolder(items: List<BaseItem>, position: Int, holder: RecyclerView.ViewHolder, payloads: MutableList<Any>) {
        (holder as ViewHolder).bind(items[position] as EpisodeOptionsItem)
    }

    @SuppressLint("ResourceType")
    internal inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        init {
            itemView.constraint.background = itemView.context.themeDrawable(R.drawable.card, R.attr.colorBackgroundWindow)
            val icon = itemView.context.themeDrawable(R.drawable.ic_play_circle_filled, R.attr.colorText)
            itemView.continueButton.setCompoundDrawablesWithIntrinsicBounds(icon, null, null, null)
        }

        fun bind(optionsItem: EpisodeOptionsItem) {
            with(itemView) {
                val buttonText: String = if (optionsItem.isFirst) {
                    itemView.resources.getString(R.string.watch_now)
                } else {
                    itemView.resources.getString(R.string.watch_continue)
                }
                episodeTextView.text = optionsItem.episodeItem.episodeFull
                continueButton.text = buttonText
                continueButton.setOnClickListener { callback.invoke(EpisodeOptionAction.ALTERNATIVE_SOURCE, optionsItem.episodeItem) }
                alternativeImageView.visibility = View.GONE
            }
        }
    }
}